### ISSUE TITLE
WAMP-cryptosign

### ISSUE THREAD

**Issue by oberstet** [wamp-proto/wamp-proto#230]
New WAMP level authentication scheme using Ed25519.

**Comment by oberstet** [comment]
A client that wants to connect to a WAMP router using WAMP-cryptosign authentiation first needs a public-private Ed25519 key pair.

With Ed25519, the private key is derived from a 32 byte random value, which when persisted, again is called private key, because the actual derivation of the private key is purely deterministic.

The public key is also derived from the private key using a deterministic scheme, and again has length 32 bytes.

To connect, the client sends a `[HELLO, Realm|uri, Details|dict]` message to the router.

The `Realm|uri` can also be `null` to request dynamic realm assignment by the router. The `Details|dict` must have the following attributes:

- `auth-methods`: A **list of strings** with authentication methods the client is ready to perform, in descending preferrence, eg `["cryptosign"]`
- `auth-id`: A **string** with the base64 encoding of the client public key
- `auth-request-serializer`: A **string** with one of these values `"json"`, `"msgpack"` or `"cbor"`
- `auth-request`: A **string**, see below.

The `auth-request` is constructed as follows. A **dict** with the following attributes is created:

- `router-pubkey`: A **string** with the base64 encoding of the router public key the client wishes to connect.
- `client-machine-id`: 
- `client-time`: A **string** with the client time in UTC, and formatted according to ISO-8601, e.g. `"2016-01-01T15:26:36.809Z"`
- `client-nonce`: A **string** with the base64 encoding of a 32 byte random nonce value.

The dict is then serialized according to the `auth-request-serializer` specified, and the binary string resulting is base64 encoded.

* `[CHALLENGE, AuthMethod|string, Extra|dict]`
* `[AUTHENTICATE, Signature|string, Extra|dict]`
* `[WELCOME, Session|id, Details|dict]`
* `[ABORT, Details|dict, Reason|uri]`

**Comment by oberstet** [comment]
# CHALLENGE message

A router receiving a `HELLO` message with a request to perform WAMP-cryptosign authentication then performs the following steps:

1. read and decode the client public key from the `auth-id`
2. check the `auth-request-signature`
3. if the signature is valid, deserialize `auth-request` according to `auth-request-serializer`
4. lookup the router private that matches the given `router-pubkey`
5. generate a 32 byte random nonce

The router must then send a  `[CHALLENGE, AuthMethod|string, Extra|dict]` message to the client, with `AuthMethod = "cryptosign"` and `Extra|dict` having the following attributes:

- `challenge`: A **string**, see below.
- `challenge-signature`: The base64 encoding of the client's Ed25519 signature (a 64 byte value) on the `challenge` data.

The `challenge` is constructed as follows. A **dict** with the following attributes is created:

- `client-nonce`: 
- `router-nonce`: 
- `router-time` (optional): 
- `router-machine-id` (optional): 

The dict is then serialized according to the `auth-request-serializer`, and the binary string resulting is base64 encoded in `challenge`.

**Comment by oberstet** [comment]
# AUTHENTICATE message

When a client receives a `CHALLENGE` message from the router, it performs the following steps:

1. check the `challenge-signature` versus the router public key
2. if the signature is valid, deserialize `challenge` using the same serializer originally sent in `HELLO`
3. verify that the `client-nonce` matches the one originally sent in `HELLO`
4. sign the `challenge` using the client private key

The Ed25519 signature created (a 64 byte value) is then base64 encoded, and a `[AUTHENTICATE, Signature|string, Extra|dict]` message is sent to the router. The `Extra|dict` is the empty dict.

**Comment by oberstet** [comment]
# WELCOME message

When a router receives a `AUTHENTICATE` message from the client, it performs the following steps:

1. check that `AUTHENTICATE.Signature` matches the expected one versus the client public key
2. if the client has requested an explicit realm, check if the realm exists and the client is allowed to join using the client public key as client identifier
3. if the client hasn't specified a realm, lookup the default realm for the client based on the client public key as client identifier

Send a `[WELCOME, Session|id, Details|dict]` message to the client.

If the realm was assigned dynamically, the `Details.realm|string` attribute MUST be present.

**Comment by oberstet** [comment]
Some words regarding certificates. WAMP-cryptosign deliberately leaves out certificates, and "only" retains the core of a PKI. The management, distribution and possibly revocation of public keys as well as private keys, if there is any sharing of private keys across endpoints, is left to the application.

**Comment by oberstet** [comment]
Here is a log of a `wamp-cryptosign` authentication handshake (only the client is authenticated, and the challenge is much simplified):

```console
(python279_1)oberstet@thinkpad-t430s:~/scm/crossbario/crossbarexamples/authentication/cryptosign/static$ python client.py --realm devices --authid client01@example.com --key client01.key
pubkey = 545efb0a2192db8d43f118e9bf9aee081466e1ef36c708b96ee6f62dddad9122
Connecting to ws://localhost:8080/ws: realm=devices, authid=client01@example.com
2016-01-05T00:53:30+0100 ClientSession connected. Joining realm <devices> under authid <client01@example.com>
2016-01-05T00:53:30+0100 TX WAMP HELLO Message (realm = devices, roles = {u'subscriber': subscriber(publisher_identification=True, payload_transparency=True, pattern_based_subscription=True, subscription_revocation=True, payload_encryption_cryptobox=True), u'publisher': publisher(publisher_identification=True, publisher_exclusion=True, payload_transparency=True, subscriber_blackwhite_listing=True, payload_encryption_cryptobox=True), u'caller': caller(payload_encryption_cryptobox=True, progressive_call_results=True, payload_transparency=True, caller_identification=True), u'callee': callee(payload_encryption_cryptobox=True, payload_transparency=True, pattern_based_registration=True, shared_registration=True, caller_identification=True, registration_revocation=True, progressive_call_results=True)}, authmethods = [u'cryptosign'], authid = client01@example.com, authrole = None)
2016-01-05T00:53:30+0100 RX WAMP CHALLENGE Message (method = cryptosign, extra = {u'challenge': 'fb645e01a5c3e54a718b5e1f91e887bdb16290e1088fc5b6e2b5230e28d690e4'})
2016-01-05T00:53:30+0100 ClientSession challenge received: Challenge(method=cryptosign, extra={u'challenge': 'fb645e01a5c3e54a718b5e1f91e887bdb16290e1088fc5b6e2b5230e28d690e4'})
2016-01-05T00:53:30+0100 TX WAMP AUTHENTICATE Message (signature = 5f336d51e8d058c08b754ad81ce76e55c4f13d0ab17f29a3f727e65c0b061ad33e47e169a0143569d5855c5eb1130e015f4ed716067599da05c60aec75e2ba06fb645e01a5c3e54a718b5e1f91e887bdb16290e1088fc5b6e2b5230e28d690e4, extra = {})
2016-01-05T00:53:30+0100 RX WAMP WELCOME Message (session = 6305273498872453, roles = {u'broker': broker(publisher_identification=True, pattern_based_subscription=True, subscription_meta_api=True, payload_encryption_cryptobox=True, payload_transparency=True, subscriber_blackwhite_listing=True, session_meta_api=True, publisher_exclusion=True, subscription_revocation=True), u'dealer': dealer(payload_encryption_cryptobox=True, payload_transparency=True, pattern_based_registration=True, registration_meta_api=True, shared_registration=True, caller_identification=True, session_meta_api=True, registration_revocation=True, progressive_call_results=True)}, realm = devices, authid = client01@example.com, authrole = device, authmethod = cryptosign, authprovider = static, authextra = None)
2016-01-05T00:53:30+0100 ClientSession joined: SessionDetails(realm=<devices>, session=6305273498872453, authid=<client01@example.com>, authrole=<device>, authmethod=cryptosign, authprovider=static, authextra=None)
2016-01-05T00:53:30+0100 TX WAMP GOODBYE Message (message = None, reason = wamp.close.normal)
2016-01-05T00:53:30+0100 RX WAMP GOODBYE Message (message = None, reason = wamp.close.normal)
2016-01-05T00:53:30+0100 ClientSession left: CloseDetails(reason=<wamp.close.normal>, message='None')
2016-01-05T00:53:30+0100 WAMP-over-WebSocket transport lost: wasClean = True, code = 1000, reason = 'None'
2016-01-05T00:53:30+0100 ClientSession disconnected
2016-01-05T00:53:30+0100 Main loop terminated.
```

**Comment by davidwdan** [comment]
Is there an updated spec for cryosign?  This one looks like it's out of date.

**Comment by oberstet** [comment]
@davidwdan yeah, the current text needs some work. some details, but also stuff like TLS channel binding need text

**Comment by oberstet** [comment]
This should also have an intro section that goes into the background, differences and use for common elliptic curves:

* secp256k1: blockchains 
* secp256r1 (== NIST P-256, =prime256v1): TLS
* curve25519: upcoming

=> https://safecurves.cr.yp.to/

**Comment by konsultaner** [comment]
To @davidwdan and all other implementers, You can find an additional (dartlang) implementation of cryptosign here:

https://github.com/konsultaner/connectanum-dart/blob/8ec1c5f10dd9952aa839e2347f9053796ddbde37/lib/src/authentication/cryptosign_authentication.dart
